### PR TITLE
Pass along sharedData from consent response to 3p ad iframes

### LINKED ISSUE #14921: amp-consent: forward KV pairs from consent remote endpoint to 3p vendors
There are use cases where the remote endpoint want to share more information in addition to the consent state to 3p vendors, such as if the current user is under certain of age or so.

We propose a generic data object named `sharedData` in the checkConsentHRref response, which will be passed along with consent state to any 3p vendors that are interested in consent state.

```
{
  promptIfUnknown: {boolean}
  sharedData: {Object}
}
```

For privacy reason, this `sharedData` is not persisted in client side storage. It will be fetched and forwarded on page view basis. 

To get this data, 3p vendor extensions can call
```
getConsentPolicySharedData(policyId)
```

Master issue: #13716

### PR DESCRIPTION
**Pull request by lannka**
Closes #14921

### CODE REVIEW

File: extensions/amp-consent/0.1/amp-consent.js [hunk 1]
```diff
@@ -286,10 +286,15 @@ export class AmpConsent extends AMP.BaseElement {
...
-              this.parseConsentResponse_(instanceId, response);
-            });
+        const remoteConsentPromise = this.getConsentRemote_(instanceId);
+        isConsentRequiredPromise = remoteConsentPromise.then(response => {
+          return this.parseConsentResponse_(instanceId, response);
+        });
+        const sharedDataPromise = remoteConsentPromise.then(response => {
+          return response ? response['sharedDataPromise'] : null;
+        });
+        this.consentStateManager_.setConsentInstanceSharedData(
```
Comment (zhouyx) [inline extensions/amp-consent/0.1/amp-consent.js]: why we pass the sharedDataPromise here, but not just sharedData?

Comment (lannka) [reply extensions/amp-consent/0.1/amp-consent.js]: This is to completely avoid risks of race condition later on.

File: build-system/app.js [hunk 2]
```diff
@@ -552,6 +552,10 @@ app.post('/get-consent-v1/', (req, res) => {
   assertCors(req, res, ['POST']);
   const body = {
     'promptIfUnknown': true,
+    'sharedDataPromise': {
```
Comment (zhouyx) [inline build-system/app.js]: `sharedDataPromise`? I thought we name it `sharedData`?

File: extensions/amp-consent/0.1/consent-policy-manager.js [hunk 3]
```diff
@@ -155,6 +177,11 @@ export class ConsentPolicyInstance {
     }
   }
 
+  /** @returns {Array} */
```
Comment (zhouyx) [inline extensions/amp-consent/0.1/consent-policy-manager.js]: `{!Array<string>}`

File: extensions/amp-consent/0.1/consent-state-manager.js [hunk 4]
```diff
@@ -112,6 +112,33 @@ export class ConsentStateManager {
     return unlistener;
   }
 
+
+  /**
+   * Sets a promise which resolves to a shareData object that is to be returned
+   * from the remote endpoint.
+   *
+   * @param instanceId
```
Comment (zhouyx) [inline extensions/amp-consent/0.1/consent-state-manager.js]: nit {string} and {?Object}

File: extensions/amp-consent/0.1/consent-policy-manager.js [hunk 5]
```diff
@@ -109,6 +109,28 @@ export class ConsentPolicyManager {
...
+        .then(manager => {
+          const promises = this.instances_[policyId].getConsentInstanceIds()
+              .map(consentId =>
+                manager.getConsentInstanceSharedData(consentId));
+          return Promise.all(promises);
+        }).then(sharedDatas => {
+          // preprend an empty object
+          // since Object.assign does not accept null as first argument
+          sharedDatas.unshift({});
+          return Object.assign.apply(null, sharedDatas);
```
Comment (zhouyx) [inline extensions/amp-consent/0.1/consent-policy-manager.js]: don't really care, but please comment that key pairs in sharedData may be overwritten after merge. We can come back to this once multi consents is supported.

File: extensions/amp-consent/0.1/amp-consent.js (hunk 1 above)
Comment (zhouyx) [reply extensions/amp-consent/0.1/amp-consent.js]: Actually if we are going to support suppress consent state, we will need to wait for response from remote endpoint before resolve any consent policy. In that case there won't be any race condition. 
But if you want to add an extra layer of protection, this also looks good.

File: build-system/app.js (hunk 2 above)
Comment (lannka) [reply build-system/app.js]: good catch. blame my IDE :-)

File: extensions/amp-consent/0.1/amp-consent.js (hunk 1 above)
Comment (lannka) [reply extensions/amp-consent/0.1/amp-consent.js]: Yep. I think this is a safer way to do it.

File: extensions/amp-consent/0.1/consent-policy-manager.js (hunk 5 above)
Comment (lannka) [reply extensions/amp-consent/0.1/consent-policy-manager.js]: expaned the js doc to talk about how the data gets merged

File: extensions/amp-consent/0.1/consent-state-manager.js (hunk 4 above)
Comment (lannka) [reply extensions/amp-consent/0.1/consent-state-manager.js]: done

File: extensions/amp-consent/0.1/consent-policy-manager.js (hunk 3 above)
Comment (lannka) [reply extensions/amp-consent/0.1/consent-policy-manager.js]: done

File: extensions/amp-consent/0.1/amp-consent.js [hunk 6]
```diff
@@ -286,10 +286,15 @@ export class AmpConsent extends AMP.BaseElement {
...
-        isConsentRequiredPromise = this.getConsentRemote_(instanceId).then(
-            response => {
-              this.parseConsentResponse_(instanceId, response);
-            });
+        const remoteConsentPromise = this.getConsentRemote_(instanceId);
+        isConsentRequiredPromise = remoteConsentPromise.then(response => {
+          return this.parseConsentResponse_(instanceId, response);
+        });
+        const sharedDataPromise = remoteConsentPromise.then(response => {
+          return response ? response['sharedData'] : null;
```
Comment (zhouyx) [inline extensions/amp-consent/0.1/amp-consent.js]: Looks like the promise can return `undefined` here.

File: extensions/amp-consent/0.1/consent-policy-manager.js (hunk 5 above)
Comment (zhouyx) [reply extensions/amp-consent/0.1/consent-policy-manager.js]: Do you plan to ask individual ad network to add their sharedData info to doc?

File: extensions/amp-consent/0.1/amp-consent.js (hunk 6 above)
Comment (lannka) [reply extensions/amp-consent/0.1/amp-consent.js]: should be fine right? it can be `undefined` or `null`.

Comment (lannka) [reply extensions/amp-consent/0.1/amp-consent.js]: done

File: extensions/amp-consent/0.1/consent-policy-manager.js (hunk 5 above)
Comment (lannka) [reply extensions/amp-consent/0.1/consent-policy-manager.js]: yes. document will come later.

File: extensions/amp-consent/0.1/consent-state-manager.js [hunk 7]
```diff
@@ -112,6 +112,33 @@ export class ConsentStateManager {
...
+        `${TAG}: cannot find this instance`);
+    this.instances_[instanceId].sharedDataPromise = sharedDataPromise;
+  }
+
+  /**
+   * Returns a promise that resolves to a shareData object that is returned
+   * from the remote endpoint.
+   *
+   * @param {string} instanceId
+   * @return {?Promise<?Object>}
```
Comment (zhouyx) [inline extensions/amp-consent/0.1/consent-state-manager.js]: need to include undefined here as well

**Pull request merged**